argutils/ancestry.py

```diff
@@ -572,3 +573,127 @@ def ca(*args):
+
+    def contains(self, x):
+        for left, right in self.intervals:
+            if left <= x < right:
+                return True
+        return False
+
+    def union(self, other):
+        """
+        Returns a new IntervalSet with the union of intervals in this and
+        other.
+        """
+        # Dreadful implementation, but oh well.
+        intervals = self.intervals.copy()
+        for interval in other.intervals:
+            if interval not in intervals:
+                intervals.append(interval)
+        return IntervalSet(intervals)
+
+
+import intervaltree
+
+def as_garg(ts):
+    """
+    Returns the specified unresolved ARG as an GARG E
+    """
+    E = [(edge.child, edge.parent, intervaltree.IntervalTree.from_tuples([(edge.left, edge.right)]))
+        for edge in ts.edges()]
+    return E
+
+def as_resolved_garg(ts):
+    """
+    Returns the specified unresolved ARG as an GARG E with respect to
+    the tree sequences samples.
+    """
+    S = ts.samples()
+    E = as_garg(ts)
+    print(E)
+    I = collections.defaultdict(intervaltree.IntervalTree)
+    for u in S:
+        I[u] = intervaltree.IntervalTree.from_tuples([(0, ts.sequence_length)])
+
+    N = set(S)
+    while len(N) > 0:
+        c = N.pop()
+        for e in [e for e in E if e[0] == c]:
+            p = e[1]
+            print("parent = ", p)
+            print("Ip = ", I[p])
+            print("Ic = ", I[c])
+            print("e = ", e[2])
+            print("intersection = ", I[c].intersection(e[2]))
+            I[p] = I[p].union(I[c].intersection(e[2]))
+            N.add(p)
+    return [(c, p, I[c]) for c, p, _ in E]
+
+def as_earg(ts):
+    """
+    Returns the specified unresolved ARG as an EARG (E, sigma).
+    """
+    sigma = np.full(ts.num_nodes, int(ts.sequence_length), dtype=int)
+    edges = iter(ts.edges())
+    edge = next(edges, None)
+    E = []
+    while edge is not None:
+        if edge.left == 0 and edge.right == ts.sequence_length:
+            E.append((edge.child, edge.parent))
+        else:
+            assert edge.left == 0
+            breakpoint = edge.right
+            child = edge.child
+            sigma[child] = breakpoint
+            E.append((edge.child, edge.parent))
+            edge = next(edges, None)
+            assert edge.left == breakpoint
+            assert edge.child == child
+            E.append((edge.child, edge.parent))
+        edge = next(edges, None)
+    # We could wrap sigma as a function to be literal about the definition
+    # but this is simpler for debugging.
+    return E, sigma
+
+
+def earg_get_tree(E, sigma, S, x):
+    """
+    Given a minimal EARG definition (E, sigma), return the tree for a the given
+    set of samples at the specified position as a dictionary parent->child.
+    """
+    N = set(S)
+    parent = {}
+    while len(N) > 0:
+        c = N.pop()
+        P = [e[1] for e in E if e[0] == c]
+        if len(P) > 0:
+            if x < sigma[c]:
```

Comment:
Note: this is why I like defining sigma(u) = inf for non recombination nodes. This is about as cleanly as you can implement this operation.

Reply:
I can see why, computationally, you might want to use `inf`, but that's not the same as using it in the mathematical definition, is it?

Reply:
Yan beat me to it, but I was going to make the same point. I'm not strongly opposed to using inf in the text, but it always stands out to me as a bit odd when looking through the draft.

Reply:
Let's see how it goes - I have some draft text describing how we build the trees in the text, and it does help make that description simpler and more concise. We may or may not want to include this. Maybe it's *too* neat though! Anyway, that's my motivation - let's leave it for now.